benchmarks/core/benchmark_copy.cpp

```diff
@@ -0,0 +1,118 @@
+/*
+ * VecMem project, part of the ACTS project (R&D line)
+ *
+ * (c) 2022 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+// VecMem include(s).
+#include <vecmem/memory/host_memory_resource.hpp>
+#include <vecmem/utils/copy.hpp>
+
+// Common benchmark include(s).
+#include "../common/make_jagged_sizes.hpp"
+#include "../common/make_jagged_vector.hpp"
+
+// Google benchmark include(s).
+#include <benchmark/benchmark.h>
+
+// System include(s).
+#include <vector>
+
+namespace vecmem::benchmark {
+
+/// The (host) memory resource to use in the benchmark(s).
+static host_memory_resource host_mr;
+/// The copy object to use in the benchmark(s).
+static copy host_copy;
+
+/// Function benchmarking "unknown" host-to-device jagged vector copies
+void jaggedVectorUnknownHtoDCopy(::benchmark::State& state) {
+
+    // Generate the sizes of the jagged vector/buffer for the test.
+    const std::vector<std::size_t> sizes =
+        make_jagged_sizes(state.range(0), state.range(1));
```

Comment:
This goes for all these benchmarks: non-deterministically generating the sizes once and then repeatedly benchmarking them is not great benchmark design. Ideally you would regenerate these sizes for each iteration of the benchmark, or generate them deterministically.

Reply:
:thinking: For a given "measurement point" I don't want to vary the buffer sizes. As that would require us to re-allocate the buffers all the time. And would have to instruct Google Benchmark to ignore the time taken by that operation. (Since I'm not interested in benchmarking the allocations here.)

The non-determinism is not great, I agree though. Since the exact measurements become dependent on the order in which the measurements are made. Which did cause me issues while trying to run individual tests one-by-one. (During debugging.)

I could set the seed of the random number generation explicitly... :thinking:

Reply:
Agreed, setting the seed explicitly would be good. Or, given some strided parameter _s_ and the given maximum _m_, generate the sequence [0, _s_ mod _m_, 2 _s_ mod _m_, 3 _s_ mod _m_, ...].

Reply:
I really don't want to overthink these benchmarks. I only created them to see if my changes would help at all. I don't want to make these into anything exact. The traccc code will be what should tell us how well the copies are behaving. By comparing what happens when we include these changes, and we start specifying `vecmem::copy::type::` parameters to the jagged vector copies...

Reply:
Okay, fine, but in that case please do add a custom counter for the number of bytes transferred (and, ideally, the rate (= bandwidth)), which should literally be a one liner. https://github.com/google/benchmark/blob/main/docs/user_guide.md#custom-counters

Reply:
This was a good idea. I quite like the output from Google Benchmark.

```
[bash][atspot01]:build > ./bin/vecmem_benchmark_cuda --benchmark_filter="jaggedVector"
2022-06-03T13:57:26+02:00
Running ./bin/vecmem_benchmark_cuda
Run on (16 X 3901.36 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x8)
  L1 Instruction 32 KiB (x8)
  L2 Unified 256 KiB (x8)
  L3 Unified 16384 KiB (x1)
Load Average: 0.66, 0.92, 1.31
--------------------------------------------------------------------------------------------------
Benchmark                                        Time             CPU   Iterations UserCounters...
--------------------------------------------------------------------------------------------------
jaggedVectorUnknownHtoDCopy/10/50            30321 ns        30321 ns        23403 Bytes=1056 Rate=34.8272M/s
jaggedVectorUnknownHtoDCopy/64/50           204871 ns       204870 ns         3428 Bytes=6.428k Rate=31.3761M/s
jaggedVectorUnknownHtoDCopy/512/50         1667556 ns      1667557 ns          417 Bytes=51.304k Rate=30.766M/s
jaggedVectorUnknownHtoDCopy/4096/50       13469321 ns     13469167 ns           52 Bytes=413.384k Rate=30.6911M/s
jaggedVectorUnknownHtoDCopy/32768/50     107251697 ns    107248511 ns            6 Bytes=3.25954M Rate=30.3924M/s
jaggedVectorUnknownHtoDCopy/100000/50    327785463 ns    327781666 ns            2 Bytes=9.97561M Rate=30.4337M/s
jaggedVectorUnknownHtoDCopy/10/64            30022 ns        30021 ns        23144 Bytes=1.12k Rate=37.3069M/s
jaggedVectorUnknownHtoDCopy/64/64           206694 ns       206692 ns         3367 Bytes=8.612k Rate=41.666M/s
jaggedVectorUnknownHtoDCopy/512/64         1680561 ns      1680512 ns          404 Bytes=63.288k Rate=37.66M/s
jaggedVectorUnknownHtoDCopy/4096/64       13501993 ns     13502062 ns           51 Bytes=521.476k Rate=38.622M/s
jaggedVectorUnknownHtoDCopy/32768/64     107808416 ns    107808049 ns            6 Bytes=4.196M Rate=38.921M/s
jaggedVectorUnknownHtoDCopy/100000/64    329147632 ns    329144689 ns            2 Bytes=12.8249M Rate=38.9644M/s
jaggedVectorUnknownHtoDCopy/10/512           33667 ns        33665 ns        20802 Bytes=8.88k Rate=263.773M/s
jaggedVectorUnknownHtoDCopy/64/512          215378 ns       215373 ns         3230 Bytes=66.344k Rate=308.043M/s
jaggedVectorUnknownHtoDCopy/512/512        1732239 ns      1732236 ns          403 Bytes=525.596k Rate=303.421M/s
jaggedVectorUnknownHtoDCopy/4096/512      13868361 ns     13868430 ns           50 Bytes=4.17666M Rate=301.163M/s
jaggedVectorUnknownHtoDCopy/32768/512    111915517 ns    111909825 ns            6 Bytes=33.461M Rate=298.999M/s
jaggedVectorUnknownHtoDCopy/100000/512   343189369 ns    343184197 ns            2 Bytes=102.671M Rate=299.173M/s
jaggedVectorUnknownHtoDCopy/10/4096          38253 ns        38254 ns        18288 Bytes=50.312k Rate=1.31522G/s
jaggedVectorUnknownHtoDCopy/64/4096         271972 ns       271973 ns         2573 Bytes=545.992k Rate=2.00752G/s
jaggedVectorUnknownHtoDCopy/512/4096       2157755 ns      2157725 ns          323 Bytes=4.12107M Rate=1.90991G/s
jaggedVectorUnknownHtoDCopy/4096/4096     18272602 ns     18271741 ns           38 Bytes=33.8948M Rate=1.85504G/s
jaggedVectorUnknownHtoDCopy/32768/4096   147599147 ns    147594974 ns            5 Bytes=269.159M Rate=1.82363G/s
jaggedVectorUnknownHtoDCopy/100000/4096  451526029 ns    451512954 ns            2 Bytes=819.757M Rate=1.81558G/s
jaggedVectorUnknownHtoDCopy/10/5000          41248 ns        41246 ns        16742 Bytes=90.096k Rate=2.18434G/s
jaggedVectorUnknownHtoDCopy/64/5000         271284 ns       271275 ns         2575 Bytes=613.216k Rate=2.2605G/s
jaggedVectorUnknownHtoDCopy/512/5000       2230634 ns      2230501 ns          310 Bytes=5.13395M Rate=2.3017G/s
jaggedVectorUnknownHtoDCopy/4096/5000     19112686 ns     19112140 ns           37 Bytes=40.8184M Rate=2.13573G/s
jaggedVectorUnknownHtoDCopy/32768/5000   156352372 ns    156342424 ns            4 Bytes=326.085M Rate=2.08571G/s
jaggedVectorUnknownHtoDCopy/100000/5000  478669283 ns    478654574 ns            2 Bytes=1002.76M Rate=2.09496G/s
jaggedVectorKnownHtoDCopy/10/50               3231 ns         3231 ns       218419 Bytes=1056 Rate=326.813M/s
jaggedVectorKnownHtoDCopy/64/50               3942 ns         3942 ns       181199 Bytes=6.428k Rate=1.63065G/s
jaggedVectorKnownHtoDCopy/512/50             13176 ns        13175 ns        52864 Bytes=51.304k Rate=3.89398G/s
jaggedVectorKnownHtoDCopy/4096/50         13192287 ns     13191742 ns           52 Bytes=413.384k Rate=31.3366M/s
jaggedVectorKnownHtoDCopy/32768/50       106797947 ns    106793683 ns            6 Bytes=3.25954M Rate=30.5218M/s
jaggedVectorKnownHtoDCopy/100000/50        3666205 ns      3666049 ns          190 Bytes=9.97561M Rate=2.72108G/s
jaggedVectorKnownHtoDCopy/10/64               3334 ns         3334 ns       205980 Bytes=1.12k Rate=335.916M/s
jaggedVectorKnownHtoDCopy/64/64               4256 ns         4256 ns       164130 Bytes=8.612k Rate=2.02333G/s
jaggedVectorKnownHtoDCopy/512/64             14099 ns        14099 ns        49566 Bytes=63.288k Rate=4.4889G/s
jaggedVectorKnownHtoDCopy/4096/64         12520869 ns     12520621 ns           55 Bytes=521.476k Rate=41.6494M/s
jaggedVectorKnownHtoDCopy/32768/64       106655882 ns    106652525 ns          100 Bytes=4.196M Rate=39.3427M/s
jaggedVectorKnownHtoDCopy/100000/64      329028245 ns    329021407 ns          171 Bytes=12.8249M Rate=38.979M/s
jaggedVectorKnownHtoDCopy/10/512             35732 ns        35731 ns        19912 Bytes=8.88k Rate=248.521M/s
jaggedVectorKnownHtoDCopy/64/512             11519 ns        11518 ns        59281 Bytes=66.344k Rate=5.75978G/s
jaggedVectorKnownHtoDCopy/512/512            65653 ns        65652 ns        10632 Bytes=525.596k Rate=8.00585G/s
jaggedVectorKnownHtoDCopy/4096/512        13992368 ns     13992081 ns         1228 Bytes=4.17666M Rate=298.502M/s
jaggedVectorKnownHtoDCopy/32768/512        7792626 ns      7792455 ns           86 Bytes=33.461M Rate=4.29402G/s
jaggedVectorKnownHtoDCopy/100000/512      26133435 ns     26133050 ns           27 Bytes=102.671M Rate=3.92879G/s
jaggedVectorKnownHtoDCopy/10/4096            38437 ns        38437 ns        18560 Bytes=50.312k Rate=1.30895G/s
jaggedVectorKnownHtoDCopy/64/4096            67534 ns        67531 ns        10316 Bytes=545.992k Rate=8.08506G/s
jaggedVectorKnownHtoDCopy/512/4096         1936100 ns      1936067 ns          360 Bytes=4.12107M Rate=2.12858G/s
jaggedVectorKnownHtoDCopy/4096/4096       17727057 ns     17726640 ns           41 Bytes=33.8948M Rate=1.91208G/s
jaggedVectorKnownHtoDCopy/32768/4096      63160896 ns     63159817 ns           11 Bytes=269.159M Rate=4.26155G/s
jaggedVectorKnownHtoDCopy/100000/4096    382813696 ns    382788437 ns            2 Bytes=819.757M Rate=2.14154G/s
jaggedVectorKnownHtoDCopy/10/5000            14019 ns        14019 ns        49834 Bytes=90.096k Rate=6.42687G/s
jaggedVectorKnownHtoDCopy/64/5000            75168 ns        75167 ns         9297 Bytes=613.216k Rate=8.15803G/s
jaggedVectorKnownHtoDCopy/512/5000         2081627 ns      2081598 ns          337 Bytes=5.13395M Rate=2.46635G/s
jaggedVectorKnownHtoDCopy/4096/5000       17855199 ns     17855114 ns           39 Bytes=40.8184M Rate=2.28609G/s
jaggedVectorKnownHtoDCopy/32768/5000      72954777 ns     72952146 ns           10 Bytes=326.085M Rate=4.46985G/s
jaggedVectorKnownHtoDCopy/100000/5000    464232754 ns    463969856 ns            2 Bytes=1002.76M Rate=2.16126G/s
jaggedVectorUnknownDtoHCopy/10/50            37921 ns        37921 ns        18482 Bytes=1056 Rate=27.8474M/s
jaggedVectorUnknownDtoHCopy/64/50           264572 ns       264562 ns         2727 Bytes=6.428k Rate=24.2968M/s
jaggedVectorUnknownDtoHCopy/512/50         2337802 ns      2337806 ns          299 Bytes=51.304k Rate=21.9454M/s
jaggedVectorUnknownDtoHCopy/4096/50       19191487 ns     19191213 ns           37 Bytes=413.384k Rate=21.5403M/s
jaggedVectorUnknownDtoHCopy/32768/50     153579194 ns    153580038 ns            5 Bytes=3.25954M Rate=21.2237M/s
jaggedVectorUnknownDtoHCopy/100000/50    468695949 ns    468673681 ns            2 Bytes=9.97561M Rate=21.2848M/s
jaggedVectorUnknownDtoHCopy/10/64            42513 ns        42513 ns        16430 Bytes=1.12k Rate=26.3449M/s
jaggedVectorUnknownDtoHCopy/64/64           289591 ns       289583 ns         2409 Bytes=8.612k Rate=29.7393M/s
jaggedVectorUnknownDtoHCopy/512/64         2356066 ns      2356057 ns          297 Bytes=63.288k Rate=26.8618M/s
jaggedVectorUnknownDtoHCopy/4096/64       19351761 ns     19350688 ns           36 Bytes=521.476k Rate=26.9487M/s
jaggedVectorUnknownDtoHCopy/32768/64     154385051 ns    154383929 ns            5 Bytes=4.196M Rate=27.179M/s
jaggedVectorUnknownDtoHCopy/100000/64    470208851 ns    470195926 ns            2 Bytes=12.8249M Rate=27.2757M/s
jaggedVectorUnknownDtoHCopy/10/512           47350 ns        47348 ns        14781 Bytes=8.88k Rate=187.548M/s
jaggedVectorUnknownDtoHCopy/64/512          304557 ns       304555 ns         2301 Bytes=66.344k Rate=217.839M/s
jaggedVectorUnknownDtoHCopy/512/512        2439691 ns      2439572 ns          287 Bytes=525.596k Rate=215.446M/s
jaggedVectorUnknownDtoHCopy/4096/512      19932820 ns     19932941 ns           35 Bytes=4.17666M Rate=209.536M/s
jaggedVectorUnknownDtoHCopy/32768/512    160008435 ns    160006883 ns            4 Bytes=33.461M Rate=209.122M/s
jaggedVectorUnknownDtoHCopy/100000/512   488397951 ns    488391428 ns            2 Bytes=102.671M Rate=210.223M/s
jaggedVectorUnknownDtoHCopy/10/4096          53317 ns        53317 ns        13084 Bytes=50.312k Rate=943.647M/s
jaggedVectorUnknownDtoHCopy/64/4096         373096 ns       373096 ns         1873 Bytes=545.992k Rate=1.46341G/s
jaggedVectorUnknownDtoHCopy/512/4096       3026762 ns      3026656 ns          231 Bytes=4.12107M Rate=1.36159G/s
jaggedVectorUnknownDtoHCopy/4096/4096     24388963 ns     24388775 ns           29 Bytes=33.8948M Rate=1.38977G/s
jaggedVectorUnknownDtoHCopy/32768/4096   194531709 ns    194529653 ns            4 Bytes=269.159M Rate=1.38364G/s
jaggedVectorUnknownDtoHCopy/100000/4096  593695412 ns    593692090 ns            1 Bytes=819.757M Rate=1.38078G/s
jaggedVectorUnknownDtoHCopy/10/5000          59459 ns        59459 ns        11713 Bytes=90.096k Rate=1.51526G/s
jaggedVectorUnknownDtoHCopy/64/5000         384042 ns       384041 ns         1823 Bytes=613.216k Rate=1.59675G/s
jaggedVectorUnknownDtoHCopy/512/5000       3187407 ns      3187399 ns          219 Bytes=5.13395M Rate=1.6107G/s
jaggedVectorUnknownDtoHCopy/4096/5000     25465552 ns     25465238 ns           27 Bytes=40.8184M Rate=1.60291G/s
jaggedVectorUnknownDtoHCopy/32768/5000   203708089 ns    203706873 ns            3 Bytes=326.085M Rate=1.60076G/s
jaggedVectorUnknownDtoHCopy/100000/5000  622520757 ns    622497479 ns            1 Bytes=1002.76M Rate=1.61087G/s
jaggedVectorKnownDtoHCopy/10/50               5445 ns         5445 ns       127428 Bytes=1056 Rate=193.929M/s
jaggedVectorKnownDtoHCopy/64/50               7332 ns         7332 ns        95301 Bytes=6.428k Rate=876.729M/s
jaggedVectorKnownDtoHCopy/512/50             23474 ns        23473 ns        29804 Bytes=51.304k Rate=2.18564G/s
jaggedVectorKnownDtoHCopy/4096/50           166866 ns       166867 ns         4192 Bytes=413.384k Rate=2.47732G/s
jaggedVectorKnownDtoHCopy/32768/50       155796053 ns    155792092 ns            4 Bytes=3.25954M Rate=20.9223M/s
jaggedVectorKnownDtoHCopy/100000/50        3809098 ns      3809057 ns          184 Bytes=9.97561M Rate=2.61892G/s
jaggedVectorKnownDtoHCopy/10/64               5357 ns         5357 ns       130188 Bytes=1.12k Rate=209.085M/s
jaggedVectorKnownDtoHCopy/64/64               8059 ns         8059 ns        86259 Bytes=8.612k Rate=1068.58M/s
jaggedVectorKnownDtoHCopy/512/64             26873 ns        26872 ns        25899 Bytes=63.288k Rate=2.35516G/s
jaggedVectorKnownDtoHCopy/4096/64           183990 ns       183989 ns         3803 Bytes=521.476k Rate=2.83428G/s
jaggedVectorKnownDtoHCopy/32768/64       156747911 ns    156747236 ns            4 Bytes=4.196M Rate=26.7692M/s
jaggedVectorKnownDtoHCopy/100000/64        4259123 ns      4259096 ns          164 Bytes=12.8249M Rate=3.01118G/s
jaggedVectorKnownDtoHCopy/10/512              7201 ns         7201 ns        96756 Bytes=8.88k Rate=1.23324G/s
jaggedVectorKnownDtoHCopy/64/512             22131 ns        22130 ns        31613 Bytes=66.344k Rate=2.99787G/s
jaggedVectorKnownDtoHCopy/512/512           135544 ns       135543 ns         5154 Bytes=525.596k Rate=3.87771G/s
jaggedVectorKnownDtoHCopy/4096/512          881823 ns       881804 ns          792 Bytes=4.17666M Rate=4.7365G/s
jaggedVectorKnownDtoHCopy/32768/512      161789548 ns    161788957 ns            4 Bytes=33.461M Rate=206.819M/s
jaggedVectorKnownDtoHCopy/100000/512      24975748 ns     24974586 ns           28 Bytes=102.671M Rate=4.11103G/s
jaggedVectorKnownDtoHCopy/10/4096            16353 ns        16353 ns        42840 Bytes=50.312k Rate=3.07661G/s
jaggedVectorKnownDtoHCopy/64/4096           125877 ns       125877 ns         5547 Bytes=545.992k Rate=4.33751G/s
jaggedVectorKnownDtoHCopy/512/4096          793066 ns       793059 ns          881 Bytes=4.12107M Rate=5.19642G/s
jaggedVectorKnownDtoHCopy/4096/4096        7384770 ns      7384713 ns           95 Bytes=33.8948M Rate=4.58986G/s
jaggedVectorKnownDtoHCopy/32768/4096     194409374 ns    194401568 ns            4 Bytes=269.159M Rate=1.38455G/s
jaggedVectorKnownDtoHCopy/100000/4096    419768245 ns    419757128 ns            2 Bytes=819.757M Rate=1.95293G/s
jaggedVectorKnownDtoHCopy/10/5000            25151 ns        25150 ns        27817 Bytes=90.096k Rate=3.58228G/s
jaggedVectorKnownDtoHCopy/64/5000           140856 ns       140855 ns         4964 Bytes=613.216k Rate=4.35352G/s
jaggedVectorKnownDtoHCopy/512/5000          948787 ns       948773 ns          737 Bytes=5.13395M Rate=5.41114G/s
jaggedVectorKnownDtoHCopy/4096/5000        8892506 ns      8892275 ns           79 Bytes=40.8184M Rate=4.59032G/s
jaggedVectorKnownDtoHCopy/32768/5000     199481074 ns    199472808 ns            4 Bytes=326.085M Rate=1.63473G/s
jaggedVectorKnownDtoHCopy/100000/5000    512485297 ns    512480196 ns            1 Bytes=1002.76M Rate=1.95668G/s
[bash][atspot01]:build >
```

Reply:
I'll just go ahead at this point. This way I could update [traccc](https://github.com/acts-project/traccc) to a new version of this project still today, allowing @konradkusiak97 to continue with https://github.com/acts-project/traccc/pull/201. :wink: